db-tests/src/test/java/gov/nasa/jpl/aerie/database/ExternalEventTests.java

```diff
@@ -0,0 +1,2210 @@
+        insertExternalEvent(e3);
+
+        // verify expected keys
+        final var results = getDerivedEvents();
+
+        // this range should only have a single element
+        final List<DerivedEvent> expectedResults = List.of(
+            new DerivedEvent(
+                "a",
+                SOURCE_TYPE,
+                "C",
+                DERIVATION_GROUP,
+                "2024-01-01 00:30:00+00",
+                "00:20:00",
+                "{[\"2024-01-01 00:00:00+00\",\"2024-01-01 01:00:00+00\")}",
+                "2024-01-03 00:00:00+00"
+            )
+        );
+        assertEquals(expectedResults.size(), results.size());
+        assertTrue(results.containsAll(expectedResults));
+      }
+    }
+  }
+  //endregion
+
+
+  //region Constraint Tests
+  /**
+   * The focus of this class is to provide a final set of tests to test primary and significant constraints to prove
+   *    that they work and how they work. These constraints are baked into the schema, so extensive tests of foreign or
+   *    unique keys are not excluded except for common use cases/easily made mistakes, as PSQL guarantees functionality
+   *    for us.
+   * We seek to demonstrate in this class, and verify as a sanity check.
+   */
+  @Nested
+  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+  class ConstraintTests {
+
+    /**
+     * Generally, no two sources in a derivation group, regardless of a lack of overlap, can have the same valid_at. We
+     *    demonstrate this with sparse sources. In this test, we don't even require events to demonstrate this behavior.
+     */
+    @Test
+    void sameValid_at() throws SQLException {
+      // construct the sources
+      ExternalSource a = new ExternalSource(
+          "A",
+          SOURCE_TYPE,
+          DERIVATION_GROUP,
+          "2024-01-01T00:00:00Z",
+          "2024-01-01T03:00:00Z",
+          "2024-01-01T04:00:00Z",
+          CREATED_AT);
+      ExternalSource b = new ExternalSource(
+          "B",
+          SOURCE_TYPE,
+          DERIVATION_GROUP,
+          "2024-01-01T00:00:00Z",
+          "2024-01-01T01:00:00Z",
+          "2024-01-01T02:00:00Z",
+          CREATED_AT);
+
+      // create types and first source
+      insertExternalSourceType(SOURCE_TYPE);
+
+      insertDerivationGroup(DERIVATION_GROUP, SOURCE_TYPE);
+
+      insertExternalSource(a);
+
+      // second source should fail
+      final SQLException ex = assertThrows(PSQLException.class, () -> insertExternalSource(b));
+      if (!ex.getSQLState().equals("23505")
+          && !ex.getMessage().contains("duplicate key value violates unique constraint \"dg_unique_valid_at\"")) {
+        throw ex;
+      }
+    }
+
+    /**
+     * A given source can have two (or more) events occurring at the same time, with all fields except the keys
+     *    (see noDuplicateEventsInSameSource). This test demonstrates that behavior without error.
+     */
+    @Test
+    void nEventsAtSameTime() throws SQLException {
```

Comment:
This is a derivation group test, not a constraint test, and should be moved accordingly. Specifically, it is testing the following fact:

```java
   *  An event that overlaps another event in the same source does not supersede the first event.
```